applications/FluidDynamicsApplication/custom_utilities/fluid_calculation_utilities.cpp

```diff
@@ -65,6 +65,159 @@ void FluidCalculationUtilities::UpdateValue(
+    const array_1d<double, 3>& rInput,
+    const double rShapeFunctionDerivative,
+    const IndexType DirectionIndex,
+    BoundedMatrix<double, 2, 2>& rOutput)
+{
+    KRATOS_TRY
+
+    KRATOS_DEBUG_ERROR_IF(DirectionIndex > 1)
+        << "Unsupported Direction Index. DirectionIndex should be  less than 2 "
+           "[ DirectionIndex = "
+        << DirectionIndex << " ].\n";
+
+    rOutput(0, DirectionIndex) = rInput[0] * rShapeFunctionDerivative;
+    rOutput(1, DirectionIndex) = rInput[1] * rShapeFunctionDerivative;
+
+    KRATOS_CATCH("");
+}
+
+template<>
+void FluidCalculationUtilities::UpdateGradientValue(
+    const double& rInput,
+    const double rShapeFunctionDerivative,
+    const IndexType DirectionIndex,
+    array_1d<double, 3>& rOutput)
+{
+    KRATOS_TRY
+
+    KRATOS_DEBUG_ERROR_IF(DirectionIndex > 2)
+        << "Unsupported Direction Index. DirectionIndex should be  less than 3 "
+           "[ DirectionIndex = "
+        << DirectionIndex << " ].\n";
+
+    rOutput[DirectionIndex] += rInput * rShapeFunctionDerivative;
+
+    KRATOS_CATCH("");
+}
+
+template<>
+void FluidCalculationUtilities::UpdateGradientValue(
+    const double& rInput,
+    const double rShapeFunctionDerivative,
+    const IndexType DirectionIndex,
+    array_1d<double, 2>& rOutput)
+{
+    KRATOS_TRY
+
+    KRATOS_DEBUG_ERROR_IF(DirectionIndex > 1)
+        << "Unsupported Direction Index. DirectionIndex should be  less than 2 "
+           "[ DirectionIndex = "
+        << DirectionIndex << " ].\n";
+
+    rOutput[DirectionIndex] += rInput * rShapeFunctionDerivative;
+
+    KRATOS_CATCH("");
+}
+
+template<>
+void FluidCalculationUtilities::UpdateGradientValue(
+    const array_1d<double, 3>& rInput,
+    const double rShapeFunctionDerivative,
+    const IndexType DirectionIndex,
+    BoundedMatrix<double, 3, 3>& rOutput)
+{
+    KRATOS_TRY
+
+    KRATOS_DEBUG_ERROR_IF(DirectionIndex > 2)
+        << "Unsupported Direction Index. DirectionIndex should be  less than 3 "
+           "[ DirectionIndex = "
+        << DirectionIndex << " ].\n";
+
+    rOutput(0, DirectionIndex) += rInput[0] * rShapeFunctionDerivative;
+    rOutput(1, DirectionIndex) += rInput[1] * rShapeFunctionDerivative;
+    rOutput(2, DirectionIndex) += rInput[2] * rShapeFunctionDerivative;
+
+    KRATOS_CATCH("");
+}
+
+template<>
+void FluidCalculationUtilities::UpdateGradientValue(
+    const array_1d<double, 3>& rInput,
+    const double rShapeFunctionDerivative,
+    const IndexType DirectionIndex,
+    BoundedMatrix<double, 2, 2>& rOutput)
+{
+    KRATOS_TRY
+
+    KRATOS_DEBUG_ERROR_IF(DirectionIndex > 1)
+        << "Unsupported Direction Index. DirectionIndex should be  less than 2 "
+           "[ DirectionIndex = "
+        << DirectionIndex << " ].\n";
+
+    rOutput(0, DirectionIndex) += rInput[0] * rShapeFunctionDerivative;
+    rOutput(1, DirectionIndex) += rInput[1] * rShapeFunctionDerivative;
+
+    KRATOS_CATCH("");
+}
+
```

Comment:
I wild shot (maybe not possible)

Couldn't a `boost::matrix_column<BoundedMatrix<double, x, y>>` be used here instead of `BoundedMatrix<double, x, y>`? 

I understand that you need to specialize all those combinations instead of making a template instantiation over the `TOutputDataType` because partial specialization for functions in not yet allowed and you need to assign the values differently if its a matrix or a vector, but using the proxy should allow you to access the `DirectionIndex` of the matrix your are interested in and just pass `rOutput[DirectionIndex]` as your output vector, so all functions would have essentially the same code, and you just need to explicitly instantiate at the end of the file.

Reply:
got rid of these template specializations, now it is more simpler with `std::apply`